primitives/state-machine/src/overlayed_changes.rs

```diff
@@ -498,6 +508,42 @@ impl OverlayedChanges {
 		)
 	}
 
+	/// Get an iterator over all child trees that are stored in the overlay.
+	pub fn child_infos(&self) -> impl Iterator<Item=&ChildInfo> {
+		self.prospective.children_default.iter()
+			.chain(self.committed.children_default.iter())
+			.map(|(_, v)| &v.1)
+	}
+
+	/// Get an iterator over all pending and committed changes.
+	///
+	/// Supplying `None` for `child_info` will only return changes that are in the top
+	/// tree. Specifying some `child_info` will return only the changes in that
+	/// child tree.
+	pub fn changes(&self, child_info: Option<&ChildInfo>) -> impl Iterator<Item=(&StorageKey, &OverlayedValue)> {
+		let (committed, prospective) = if let Some(child_info) = child_info {
+			match child_info.child_type() {
+				ChildType::ParentKeyId => (
+					self.committed.children_default.get(child_info.storage_key()).map(|c| &c.0),
+					self.prospective.children_default.get(child_info.storage_key()).map(|c| &c.0),
+				),
+			}
+		} else {
+			(Some(&self.committed.top), Some(&self.prospective.top))
+		};
+		committed.into_iter().flatten().chain(prospective.into_iter().flatten())
+	}
+
+	/// Return a clone of the currently pending changes.
+	pub fn clone_pending(&self) -> OverlayedChangeSet {
+		self.prospective.clone()
+	}
+
+	/// Replace the currently pending changes.
+	pub fn set_pending(&mut self, pending: OverlayedChangeSet) {
```

Comment:
I wonder if it is a good idea to introduce another name for the prospective here. 

Do you think it would be changed or removed with the upcoming PR?

Reply:
You successfully identified the remaining leak in the abstraction where we still have assumptions. I consider it a place holder that goes away with the upcoming PR. It will be replaced by rolling back a transaction. I guess even now we can replace it by just calling`discard_prospective` as the `prospective` should be empty as it is committed for every call. Actually no need to clone it. I will try if that works. Then we do not need to add further functions in that PR.

Reply:
I removed the cloning completely. It appears that just resetting the `prospective` works just as fine. I removed these two new functions. Eventually this `discard_prospective` goes away and is replaced by `rollback_transaction`.

Reply:
After talking to @bkchr we came to the conclusion that we cannot make this assumption. I revert the commit. We are back to the `pending_changes()` function until storage transaction come around.